File: android/app/src/main/java/com/warelay/app/data/remote/ApiClient.kt
package com.warelay.app.data.remote

import com.warelay.app.data.model.MatchedMessage
import com.warelay.app.data.prefs.UserPreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class ApiClient(private val preferences: UserPreferences) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val jsonMedia = "application/json; charset=utf-8".toMediaType()

    data class AuthResult(val token: String, val username: String)

    data class MessagesPage(
        val messages: List<MatchedMessage>,
        val hasMore: Boolean,
        val nextCursor: String?,
    )

    data class MessageQuery(
        val q: String? = null,
        val unread: Boolean? = null,
        val starred: Boolean? = null,
        val done: Boolean? = null,
        val isGroup: Boolean? = null,
        val folder: String? = null,
        val limit: Int = 40,
        /** Last message id from previous page (exclusive). */
        val before: String? = null,
    )

    suspend fun register(username: String, password: String): AuthResult =
        auth("/auth/register", username, password)

    suspend fun login(username: String, password: String): AuthResult =
        auth("/auth/login", username, password)

    private suspend fun auth(path: String, username: String, password: String): AuthResult =
        withContext(Dispatchers.IO) {
            val body = JSONObject()
                .put("username", username)
                .put("password", password)
                .toString()
                .toRequestBody(jsonMedia)
            val req = Request.Builder()
                .url(preferences.getHostUrl() + path)
                .post(body)
                .build()
            client.newCall(req).execute().use { res ->
                val text = res.body?.string().orEmpty()
                if (!res.isSuccessful) {
                    val err = runCatching { JSONObject(text).optString("error") }.getOrNull()
                    throw ApiException(err?.ifBlank { null } ?: "HTTP ${res.code}")
                }
                val json = JSONObject(text)
                AuthResult(
                    token = json.getString("token"),
                    username = json.getJSONObject("user").getString("username"),
                )
            }
        }

    suspend fun fetchMessages(token: String, query: MessageQuery = MessageQuery()): MessagesPage =
        withContext(Dispatchers.IO) {
            val base = preferences.getHostUrl().trimEnd('/')
            val urlBuilder = "$base/messages".toHttpUrl().newBuilder()
                .addQueryParameter("limit", query.limit.toString())
            if (!query.q.isNullOrBlank()) urlBuilder.addQueryParameter("q", query.q)
            query.unread?.let { urlBuilder.addQueryParameter("unread", it.toString()) }
            query.starred?.let { urlBuilder.addQueryParameter("starred", it.toString()) }
            query.done?.let { urlBuilder.addQueryParameter("done", it.toString()) }
            query.isGroup?.let { urlBuilder.addQueryParameter("isGroup", it.toString()) }
            if (!query.folder.isNullOrBlank() && query.folder != "all") {
                urlBuilder.addQueryParameter("folder", query.folder)
            }
            if (!query.before.isNullOrBlank()) {
                urlBuilder.addQueryParameter("before", query.before)
            }

            val req = Request.Builder()
                .url(urlBuilder.build())
                .header("Authorization", "Bearer $token")
                .get()
                .build()
            client.newCall(req).execute().use { res ->
                val text = res.body?.string().orEmpty()
                if (res.code == 401) throw UnauthorizedException()
                if (!res.isSuccessful) throw ApiException("HTTP ${res.code}")
                val json = JSONObject(text)
                val arr = json.getJSONArray("messages")
                MessagesPage(
                    messages = parseMessages(arr),
                    hasMore = json.optBoolean("hasMore", false),
                    nextCursor = json.optString("nextCursor").ifBlank { null },
                )
            }
        }

    suspend fun fetchUnreadCounts(token: String): Map<String, Int> =
        withContext(Dispatchers.IO) {
            val req = Request.Builder()
                .url(preferences.getHostUrl().trimEnd('/') + "/messages/unread-counts")
                .header("Authorization", "Bearer $token")
                .get()
                .build()
            client.newCall(req).execute().use { res ->
                val text = res.body?.string().orEmpty()
                if (res.code == 401) throw UnauthorizedException()
                if (!res.isSuccessful) throw ApiException("HTTP ${res.code}")
                val obj = JSONObject(text).getJSONObject("counts")
                val keys = listOf("all", "lgw", "lhr", "ltn", "stn", "others")
                keys.associateWith { obj.optInt(it, 0) }
            }
        }

    suspend fun patchMessage(
        token: String,
        id: String,
        read: Boolean? = null,
        starred: Boolean? = null,
        done: Boolean? = null,
    ): MatchedMessage = withContext(Dispatchers.IO) {
        val bodyJson = JSONObject()
        if (read != null) bodyJson.put("read", read)
        if (starred != null) bodyJson.put("starred", starred)
        if (done != null) bodyJson.put("done", done)
        val body = bodyJson.toString().toRequestBody(jsonMedia)
        val req = Request.Builder()
            .url(preferences.getHostUrl().trimEnd('/') + "/messages/$id")
            .header("Authorization", "Bearer $token")
            .patch(body)
            .build()
        client.newCall(req).execute().use { res ->
            val text = res.body?.string().orEmpty()
            if (res.code == 401) throw UnauthorizedException()
            if (!res.isSuccessful) {
                val err = runCatching { JSONObject(text).optString("error") }.getOrNull()
                throw ApiException(err?.ifBlank { null } ?: "HTTP ${res.code}")
            }
            parseMessage(JSONObject(text).getJSONObject("message"))
        }
    }

    suspend fun registerDevice(token: String, fcmToken: String) = withContext(Dispatchers.IO) {
        val body = JSONObject()
            .put("fcmToken", fcmToken)
            .put("platform", "android")
            .toString()
            .toRequestBody(jsonMedia)
        val req = Request.Builder()
            .url(preferences.getHostUrl() + "/devices/register")
            .header("Authorization", "Bearer $token")
            .post(body)
            .build()
        client.newCall(req).execute().use { res ->
            if (res.code == 401) throw UnauthorizedException()
            if (!res.isSuccessful) {
                val text = res.body?.string().orEmpty()
                throw ApiException(text.ifBlank { "HTTP ${res.code}" })
            }
        }
    }

    companion object {
        fun parseMessages(arr: JSONArray): List<MatchedMessage> {
            val list = mutableListOf<MatchedMessage>()
            for (i in 0 until arr.length()) {
                list += parseMessage(arr.getJSONObject(i))
            }
            return list
        }

        fun parseMessage(obj: JSONObject): MatchedMessage =
            MatchedMessage(
                id = obj.optString("id"),
                messageId = obj.optString("messageId"),
                text = obj.optString("text"),
                senderPhone = optNullableString(obj, "senderPhone"),
                senderName = optNullableString(obj, "senderName"),
                chatId = obj.optString("chatId"),
                isGroup = obj.optBoolean("isGroup"),
                waLink = optNullableString(obj, "waLink"),
                matchedPattern = optNullableString(obj, "matchedPattern"),
                folder = optNullableString(obj, "folder"),
                timestamp = optNullableString(obj, "timestamp"),
                createdAt = optNullableString(obj, "createdAt"),
                readAt = optNullableString(obj, "readAt"),
                starred = obj.optBoolean("starred", false),
                done = obj.optBoolean("done", false),
            )

        /**
         * JSONObject.optString returns the literal "null" for JSON null — treat that as absent.
         */
        private fun optNullableString(obj: JSONObject, key: String): String? {
            if (!obj.has(key) || obj.isNull(key)) return null
            val value = obj.optString(key).trim()
            if (value.isEmpty() || value == "null" || value == "undefined") return null
            return value
        }
    }
}

class ApiException(message: String) : Exception(message)
class UnauthorizedException : Exception("Unauthorized")

File: ios/WaRelay/Data/Remote/ApiClient.swift
import Foundation

struct AuthResult: Sendable {
    let token: String
    let username: String
}

struct HealthResult: Sendable {
    let ok: Bool
    let whatsappStatus: String
    let whatsappOk: Bool
    let hasQr: Bool
}

struct MessagesPage: Sendable {
    let messages: [MatchedMessage]
    let hasMore: Bool
    let nextCursor: String?
}

struct MessageQuery: Sendable {
    var q: String? = nil
    var unread: Bool? = nil
    var starred: Bool? = nil
    var done: Bool? = nil
    var thumbsUp: Bool? = nil
    var isGroup: Bool? = nil
    var folder: String? = nil
    var limit: Int = 40
    var before: String? = nil
}

enum ApiError: LocalizedError {
    case http(Int, String?)
    case unauthorized
    case invalidResponse
    case message(String)

    var errorDescription: String? {
        switch self {
        case .http(let code, let msg):
            return msg?.isEmpty == false ? msg : "HTTP \(code)"
        case .unauthorized:
            return "Unauthorized"
        case .invalidResponse:
            return "Invalid response"
        case .message(let msg):
            return msg
        }
    }
}

actor ApiClient {
    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 45
        return URLSession(configuration: config)
    }()

    func login(username: String, password: String) async throws -> AuthResult {
        try await auth(path: "/auth/login", username: username, password: password)
    }

    func register(username: String, password: String) async throws -> AuthResult {
        try await auth(path: "/auth/register", username: username, password: password)
    }

    func fetchHealth() async throws -> HealthResult {
        let url = URL(string: host() + "/health")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw ApiError.invalidResponse }
        guard (200...299).contains(http.statusCode) else {
            throw ApiError.http(http.statusCode, nil)
        }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ApiError.invalidResponse
        }
        let wa = json["whatsapp"] as? [String: Any]
        let status = (wa?["status"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "unknown"
        let connected = wa?["connected"] as? Bool ?? (status == "open")
        return HealthResult(
            ok: json["ok"] as? Bool ?? true,
            whatsappStatus: status,
            whatsappOk: wa?["ok"] as? Bool ?? connected,
            hasQr: wa?["hasQr"] as? Bool ?? false
        )
    }

    func fetchMessages(token: String, query: MessageQuery = MessageQuery()) async throws -> MessagesPage {
        var components = URLComponents(string: host() + "/messages")!
        var items: [URLQueryItem] = [
            URLQueryItem(name: "limit", value: String(query.limit)),
        ]
        if let q = query.q, !q.isEmpty { items.append(URLQueryItem(name: "q", value: q)) }
        if let unread = query.unread { items.append(URLQueryItem(name: "unread", value: String(unread))) }
        if let starred = query.starred { items.append(URLQueryItem(name: "starred", value: String(starred))) }
        if let done = query.done { items.append(URLQueryItem(name: "done", value: String(done))) }
        if let thumbsUp = query.thumbsUp { items.append(URLQueryItem(name: "thumbsUp", value: String(thumbsUp))) }
        if let isGroup = query.isGroup { items.append(URLQueryItem(name: "isGroup", value: String(isGroup))) }
        if let folder = query.folder, !folder.isEmpty, folder != "all" {
            items.append(URLQueryItem(name: "folder", value: folder))
        }
        if let before = query.before, !before.isEmpty {
            items.append(URLQueryItem(name: "before", value: before))
        }
        components.queryItems = items

        let (data, response) = try await authorizedRequest(
            url: components.url!,
            method: "GET",
            token: token
        )
        try throwIfUnauthorized(response)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let arr = json["messages"] as? [[String: Any]] else {
            throw ApiError.invalidResponse
        }
        return MessagesPage(
            messages: arr.map(MatchedMessage.init(json:)),
            hasMore: json["hasMore"] as? Bool ?? false,
            nextCursor: json["nextCursor"] as? String
        )
    }

    func fetchUnreadCounts(token: String) async throws -> [String: Int] {
        let url = URL(string: host() + "/messages/unread-counts")!
        let (data, response) = try await authorizedRequest(url: url, method: "GET", token: token)
        try throwIfUnauthorized(response)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let counts = json["counts"] as? [String: Any] else {
            throw ApiError.invalidResponse
        }
        let keys = ["all", "lgw", "lhr", "ltn", "stn", "others"]
        return Dictionary(uniqueKeysWithValues: keys.map { ($0, counts[$0] as? Int ?? 0) })
    }

    /// Mark unread messages as read. Pass folder api value, or `all`/nil for every folder.
    func markAllRead(token: String, folder: String? = nil) async throws -> [String: Int] {
        let url = URL(string: host() + "/messages/mark-all-read")!
        var body: [String: Any] = [:]
        if let folder, !folder.isEmpty {
            body["folder"] = folder
        }
        let (data, response) = try await authorizedRequest(
            url: url,
            method: "POST",
            token: token,
            jsonBody: body
        )
        try throwIfFailed(response, data: data)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let counts = json["counts"] as? [String: Any] else {
            throw ApiError.invalidResponse
        }
        let keys = ["all", "lgw", "lhr", "ltn", "stn", "others"]
        return Dictionary(uniqueKeysWithValues: keys.map { ($0, counts[$0] as? Int ?? 0) })
    }

    func patchMessage(
        token: String,
        id: String,
        read: Bool? = nil,
        starred: Bool? = nil,
        done: Bool? = nil,
        thumbsUp: Bool? = nil
    ) async throws -> MatchedMessage {
        var body: [String: Any] = [:]
        if let read { body["read"] = read }
        if let starred { body["starred"] = starred }
        if let done { body["done"] = done }
        if let thumbsUp { body["thumbsUp"] = thumbsUp }
        let url = URL(string: host() + "/messages/\(id)")!
        let (data, response) = try await authorizedRequest(
            url: url,
            method: "PATCH",
            token: token,
            jsonBody: body
        )
        try throwIfFailed(response, data: data)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let message = json["message"] as? [String: Any] else {
            throw ApiError.invalidResponse
        }
        return MatchedMessage(json: message)
    }

    func registerDevice(token: String, fcmToken: String) async throws {
        let url = URL(string: host() + "/devices/register")!
        let body: [String: Any] = [
            "fcmToken": fcmToken,
            "pushToken": fcmToken,
            "platform": "ios",
        ]
        let (_, response) = try await authorizedRequest(
            url: url,
            method: "POST",
            token: token,
            jsonBody: body
        )
        try throwIfUnauthorized(response)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            throw ApiError.http(code, nil)
        }
    }

    private func auth(path: String, username: String, password: String) async throws -> AuthResult {
        let url = URL(string: host() + path)!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "username": username,
            "password": password,
        ])
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw ApiError.invalidResponse }
        if !(200...299).contains(http.statusCode) {
            let err = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String
            throw ApiError.http(http.statusCode, err)
        }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["token"] as? String,
              let user = json["user"] as? [String: Any],
              let name = user["username"] as? String else {
            throw ApiError.invalidResponse
        }
        return AuthResult(token: token, username: name)
    }

    private func authorizedRequest(
        url: URL,
        method: String,
        token: String,
        jsonBody: [String: Any]? = nil
    ) async throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if let jsonBody {
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: jsonBody)
        }
        return try await session.data(for: request)
    }

    private func throwIfUnauthorized(_ response: URLResponse) throws {
        if let http = response as? HTTPURLResponse, http.statusCode == 401 {
            throw ApiError.unauthorized
        }
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw ApiError.http(http.statusCode, nil)
        }
    }

    private func throwIfFailed(_ response: URLResponse, data: Data) throws {
        if let http = response as? HTTPURLResponse, http.statusCode == 401 {
            throw ApiError.unauthorized
        }
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            let err = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String
            throw ApiError.http(http.statusCode, err)
        }
    }

    private func host() -> String {
        UserPreferences.hostURL()
    }
}
